### ISSUE TITLE
patched font variants (bold, semibold etc) show up as duplicate in fontbook osx

### ISSUE THREAD

**Issue by alexanderjeurissen** [ryanoasis/nerd-fonts#56]
Installing multiple font variants (bold, semibold, light, etc) on OSX using fontbook results in warnings and the font variants are all in separate "font families" ..

I tried installing multiple patched fonts (both Hack and Source Code Pro) so that should rule out that a single patched font is causing the issue. 

![screen shot 2016-02-07 at 01 29 46](https://cloud.githubusercontent.com/assets/1220084/12870233/7ef46a6e-cd3b-11e5-9343-4d40191af45a.png)

As can be seen in the screenshot, multiple font families are created for the font variants while I expect a single font family named Sauce Code Pro containing all the variants. Also all the fonts are named: "NerdFontCompleteMono"

I hope we can sort this out because I really want to use the awesome glyphs in my setup

**Comment by adambiggs** [comment]
I reported this in #25 which appeared to be fixed, but I didn't actually have a chance to test it because I stopped using the install script in favour of manually installing just the font I use... Would love to see this resolved though!

**Comment by ryanoasis** [comment]
Yes this should be fixed.. thanks for the details here.

I don't have a mac to test on but perhaps I can setup a test branch where others can try it out :smile:

**Comment by thewatts** [comment]
![screen shot 2016-02-18 at 1 06 18 am](https://cloud.githubusercontent.com/assets/1141717/13136060/e2104b42-d5db-11e5-8e36-f9f6489ed6b7.png)

Looks like it's still an issue (shown after installing `master`)

**Comment by thewatts** [comment]
Which, I must say - this project is _LEGIT_. Really thankful for it!

**Comment by ryanoasis** [comment]
@thewatts, @adambiggs

> Yes this should be fixed.. thanks for the details here

Err I apologize for not making my point clear. :blush: What I meant here was that this should _get_ fixed **not** that it _already_ was fixed. i.e. Yes this is still a problem AFAIK :+1: 

I will be leaving this ticket open until someone can confirm an attempted fix :smile: 

---

@thewatts

> Which, I must say - this project is _LEGIT_. Really thankful for it!

Thanks! That means a lot

**Comment by ryanoasis** [comment]
I will be pushing up a test branch that should address this issue. Maybe even as early as tomorrow if I get time.

I will link to it on this ticket when I do.

Thanks

**Comment by eugenesvk** [comment]
@ryanoasis I think this is an issue with FontForge not giving scripts access to font style names (https://github.com/fontforge/fontforge/issues/1061) and relying on autogessing style. Hence you get very strange styles in your fonts such as `NerdFontComplete` instead of `Regular`.
The resulting **PS Font Name** (`$fontname` set to *fontname + Suffix* in your [patcher](https://github.com/ryanoasis/nerd-fonts/blob/master/font-patcher#L90)) is identical for all different styles (e.g. `SauceCodeProNerdFontComplete` for both **Regular** and **Medium** instead of `SauceCodeProNerdComplete-Regular` and `SauceCodeProNerdComplete-Medium`) and Mac throws this error.

After I've used your script to generate the required fonts the only perfect solution to the naming mess I've found was to use [FontLab](fontlab.com)
1. set a proper **Family name** (e.g. **Sauce Code ProNerdC** for the 'Complete' version) — this would be the same for all the generated fonts with a given variation of extra glyph sets
2. Auto-generate everything with 3 buttons (`Build Style Name`, `Build Names`, and `Build OpenType Names`) to get clean names
3. Propagate these changes by clicking 'Import Names' in `Additional OpenType Names`
4. Generate a clean set of files like *SauceCodeProNerdC-MediumItalic.ttf*

**Then, you'd get the following clean names** (based on Source Code Medium Italic)
Family Name: **Sauce Code ProNerdC**
Style Name: **Medium Italic**
PS Font Name (the one that was conflicting on a Mac): **SauceCodeProNerdC-MediumItalic** (it gets automatically shortened to e.g. "It" if it's >~32 symbols)
Full Name: **Sauce Code ProNerdC Medium Italic**
(OT-Specific names) Mac Name: **Sauce Code ProNerdC Medium Italic** (though I'm not sure where these OT-specific names are relevant)

No more naming conflicts, no more "NerdFontComplete" in drop-down menus in apps!
Now, this could also be accomplished in FontForge GUI, but would be much more cumbersome. When FontForge allows access by scripts to style names as well, this could all be automated in a script.

I think it might be even possible as is if you can try to guess a style name from a file name or font properties and try to trick FontForge into reading a specially constructed full name (with a hyphen separating name from style) to write the correct style name, but I don't know how FontForge autoguess is working

**Comment by eugenesvk** [comment]
Actually, was wrong about FontForge not allowing scripting access to all the names — that github ticket threw me off. The `sfnt_names` from [FF scripting commands](https://fontforge.github.io/en-US/documentation/scripting/python/) is a string that has all the **SFNT** fields (see [Name IDs](https://partners.adobe.com/public/developer/opentype/index_name.html)) and `appendSFNTName` command allows to change any of them with either a name ID listed in the link.
So it's possible to do everything inside a FontForge script without the help of any other application (and, more importantly, without having to do any manual adjustments).

**Comment by ryanoasis** [comment]
@eugenesvk Yep! I was aware of the `sfnt_names` and actually I *think* I have the changes that should solve the issue right now. Just haven't released or pushed to any branch.

Thank you for the detailed info I think it is helpful :+1: 

I think at some point adding configuration files to denote the subfont/style name might be needed. Because some fonts have it after the dash in the filename as a convention but some don't. Also some fonts have the incorrect Styles (SubFamily) when viewed in FontForge (e.g. SourceCodePro-ExtraLight.ttf) as shown below:

![selection_23_02_16_11 15 11_794x579_001](https://cloud.githubusercontent.com/assets/8083459/13257889/bd6305ea-da1e-11e5-9715-68cb727c182e.png)

WIP Changes preview:
https://gist.github.com/ryanoasis/af51f008838aaa46fb61

particularly of note: `sourceFont.appendSFNTName('English (US)', 'SubFamily', subFamily)`

**Comment by eugenesvk** [comment]
Yeah, Source Code is unfortunately a mess in this regard (I also had to change all the font names for these fonts to be correctly reflected in e.g. ST3 — didn't understand why the same name would work on Mac, but fail to work on a Windows, then realized the font names are inconsistent).
Btw, the reason why Source Code SubFamily here is "Regular" instead of "ExtraLight" is that there is a stupid old convention of limiting styles to 4 (regular, bold, italic, bold italic). So ExtraLight doesn't fit and has to be included in a separate family "Source Code ExtraLight".
**However**, the OpenType names of this font are correct (because OT got rid of that 4-style limitation), so in your scripts you'd want to **read OpenType names** (`'Preferred Styles'` and `'Preferred Family'`)

Thanks for the gist — I'm also trying to adjust exactly this so that I don't have to do any manual changes, so it's helpful to see what you've done. 

And just to reiterate — to me the default FontLab options I've described above are ideal, so I'd suggest you follow them in your scripts as well (overwriting the default Source Code mess, for example), but of course it's up to you.

**Comment by eugenesvk** [comment]
I'd also suggest in the output to write OpenType names in addition to the regular ones, as they have a priority (e.g. Mac would display OpenType names over regular ones on conflict, think same is true for Windows, but haven't checked)

A quick reference table from my script (names in comments are from FontLab)
```
sourceFont.appendSFNTName('English (US)', 'Family', "family")                          # 'Family Name', 'Menu Name'
sourceFont.appendSFNTName('English (US)', 'SubFamily', "SubFamStyle")        # 'Style Name'
sourceFont.appendSFNTName('English (US)', 'Fullname', "fullname")                  # 'Full Name'
sourceFont.appendSFNTName('English (US)', 'Preferred Family', "PrefFam")      # OT Family Name
sourceFont.appendSFNTName('English (US)', 'Preferred Styles', "PrefSubFam") # 'OT Style Name'
sourceFont.appendSFNTName('English (US)', 'Compatible Full', "CompatFull")   # 'FOND name', (OT-specific) 'Mac Name'
```
And just FYI similar reference re. default naming functions (as far as i understand, they propagate to multiple fields in the absence of SFNT instructions, but I maybe wrong on this one)
```
familyname   # 'Family Name', 'Menu Name', (OT-specific) 'OT Family Name'
fullname       # 'Full Name', 'FOND Name', (OT-specific) 'Mac Name'
fontname     # 'PS Font Name'
```

**Comment by ryanoasis** [comment]
> old convention of limiting styles

Did *not* know this thanks!

> read OpenType names [...] And only if OpenType names aren't present, use the regular ones.

Nice. I think I even overlooked that in the screenshot I posted :blush: . Yes I think this makes sense and I this seems like the logic to go with here.

> FontLab options

Sorry I did read your suggestion but have not really gone through it in detail. Just FYI the gist I posted was work done **before** any of your messages so the new information you have provided **will** help me greatly in making this patcher 'do the right thing'. :+1: 

> shortening font names always

Also as you mentioned in #58, this one is the only one I am still not sure on. I will reply there

**Comment by eugenesvk** [comment]
> old convention of limiting styles

That's what I got from that [Adobe document](https://partners.adobe.com/public/developer/opentype/index_name.html)

> "16 Preferred Family; For historical reasons, font families have contained a maximum of four styles, but font designers may group more than four fonts to a single family. The Preferred Family allows font designers to include the preferred family grouping which contains more than four fonts. This ID is only present if it is different from ID 1."


> have not really gone through it in detail 

I'm basically arguing for three key principles: 

1. Putting full style (weight, width, and angle) in style name (e.g. **Condensed Bold Italic**) instead of only doing this for OpenType-specific names and following the useless traditions in regular names. This also means that no style information is ever contained in the 'Family Name'
2. Making OpenType-specific names and regular names identical
3. Filling in all available naming fields so that you can control the consistency of names along those unnecessary complexity of legacy duplicates

Re. https://github.com/ryanoasis/nerd-fonts/issues/58 — while thinking about shorter names, don't forget about the larger point of reducing ~36 combinations to 2 :)

**Comment by ryanoasis** [comment]
Good stuff thanks again.

Yeah I think I am in agreement with you on all 3 of the main points as you described them.

#58 - yeah that is definitely something to consider. Also claiming a bunch of font variations is a bit deceiving if most of them aren't very useful :stuck_out_tongue_closed_eyes:

**Comment by jrolfs** [comment]
Just noticed this thread after I created #61, hopefully it's a start towards solving this?

**Comment by ryanoasis** [comment]
@jrolfs Yes and I replied on #61, seems like your work at least will help solve. I might have a bit of conflict with https://gist.github.com/ryanoasis/af51f008838aaa46fb61 but I think I soon need to setup a test branch for us to check how the fonts come out.

**Comment by adamclaxon** [comment]
I have run into a variant of this issue as well. The script seems to assume the font name is the same as the family name. While the font's I have generated are now organized correctly in Font Book, all variants are generated with the same font name, and Font Book interprets them as Duplicates. The proper method I have found (at least on OS X) is for the style to be included in the font name ("OperatorMonoItalic", for instance).

**Comment by ryanoasis** [comment]
@adamclaxon 

Thanks I think we are getting close to getting this fixed now. You are correct it seems that the font name needs the style to identify it properly as a separate font variant.

--

If you want (and it would be a great help!) you can try patching with the [0.7.0 branch](https://github.com/ryanoasis/nerd-fonts/tree/0.7.0) because it seems to be solving most of the issues so far at least.

**Comment by Undistraction** [comment]
Similar / related issue:

Fontname: Example-ReguarItalic
Fontfamily: Example
Name For Humans: ExampleRegularItalic
Weight: Regular

When I generate an .otf file and open it in Fontbook, the font is shown within Example as 'Regular' rather than RegularItalic, meaning it conflicts with the genuine Regular (non-italic) face.

**Comment by mike3run** [comment]
I'm having similar issues still in the sense that I have several weights of fonts with italics versions however all of the italics get assigned the Bold weight for some reason

**Comment by ryanoasis** [comment]
@Undistraction Sorry that I never replied to you. I will test the scenario you mentioned

@mike3run Can you tell me which fonts you have patched?

**Comment by mike3run** [comment]
@ryanoasis the Operator Mono font family, so I cannot directly share it but here let me show you a screen capture of the folder with all the patched fonts and on my fontbook program you can see that for example today only the light italic version is detected (up until yesterday it had been the bold version)

![screen shot 2017-02-26 at 11 16 48](https://cloud.githubusercontent.com/assets/7088511/23341862/3d5ae628-fc15-11e6-8253-929e7bdf2055.png)

**Comment by ryanoasis** [comment]
@mike3run Thanks. Yep no worries I know about that font and yes we will comply with licenses here :smile:

**Comment by mike3run** [comment]
hey there @ryanoasis just wondering if by any chance this got solved?

Thanks a lot!

**Comment by ryanoasis** [comment]
@mike3run Hey just tested this.. so unfortunately I cannot test on macOS with fontbook but seems to me it might use the weight to differentiate the fonts.

Are you seeing the same issue with unpatched versions of the same font and same weights or does it only occur after patching?

---

Here is what I see on Linux and it seems to differentiate fine, however there does seem to be something going on:

![screenshot from 2017-04-19 19-48-35](https://cloud.githubusercontent.com/assets/8083459/25207102/3bfe5140-253a-11e7-93f2-d90c2609e037.png)

**Comment by mike3run** [comment]
@ryanoasis its a bit weird since i mean the font gets actually rendered and you can _select_ it but it will look like other weight.

So even if it says its the light font it will show as bold 🤔

**Comment by mike3run** [comment]
also forgot to add, yeah the bug happens only after patching

**Comment by ryanoasis** [comment]
@mike3run Oh! So it's not an issue with some showing up or not showing up as selectable but how the style is rendered?

I wonder if the same thing would happen for you with for instance with Hasklug (Hasklig). It seems to render the correct style/weight for me on Linux:

![selection_037](https://cloud.githubusercontent.com/assets/8083459/25207576/b2d135f0-253d-11e7-9fc9-c44c3f4d5b1e.png)

**Comment by mike3run** [comment]
Is a mixture of both, on the official fontbook app they do show with names but in a more specialized app like the one on my screenshot above or below they do not show possibly since they are being repeated:

![](https://cloud.githubusercontent.com/assets/7088511/23341862/3d5ae628-fc15-11e6-8253-929e7bdf2055.png)

**Comment by mike3run** [comment]
So you can see on the right I have all the fonts + italic versions but when imported on the italic ones only the light shows up.

On my laptop the same installed fonts make only the bold one show up as italic

**Comment by ryanoasis** [comment]
Thanks, this might take some more digging :worried:

**Comment by zxaos** [comment]
If your system is affected by this, note that having the fonts that Font Book flags as duplicates (i.e.more than a single Italic variant) active will crash iterm2 if you try to set any font from the family in use.  Disabling all italic variants but one will resolve the issue.

**Comment by mike3run** [comment]
But then you wouldn't have bold support :(((((((((((
Although yeah I actually did what you say a year ago and it works

**Comment by zxaos** [comment]
I'm not _happy_ without bold italic support, but I'd rather have the icons at all than it.

... But as this issue is happening on 1.2.0 still, it might need to be reopened? ( @ryanoasis ? )

**Comment by ryanoasis** [comment]
Maybe we can make some traction on this again this month. I won't be able to test any potential fixes but I could upload zips if others (who are on macos) can test them out?

**Comment by jackcogdill** [comment]
+1 this issue is still happening.

**Comment by MarioRicalde** [comment]
Here's the manual solution to the problem: https://github.com/ryanoasis/nerd-fonts/issues/257#issuecomment-661153972

@ryanoasis I think it's just a matter of having better Fontname generators that output shorter names.

**Comment by ryanoasis** [comment]
I know this has fallen off the radar. Trying to come up with better system to keep up with the issues

**Comment by Finii** [comment]
This is implemented/solved since `v2.2.0` via `--makegroups` command line option.
With `v3.0.0` all prepatched fonts will switch their naming to this.
Thanks for the excellent writeup, that I just now found :roll_eyes:

**Comment by zxaos** [comment]
`--makegroups` improves, but does not totally solve this problem, at least for me.

For example Operator Mono normally comes in 8 styles that FontBook recognizes as a single font:
* Bold
* Bold Italic
* Book
* Book Italic
* Light
* Light Italic
* Medium
* Medium Italic

Patching it with `--makegroups` does indeed prevent macos Fontbook from bailing on the duplicated italic names... but it also causes the `Book` and `Book Italic` styles to be split into an entirely different font (OperatorMonoSSm Book Nerd Font, vs OperatorMonoSSm Nerd Font).

Could this be reopened since it's still not fixed?

**Comment by Finii** [comment]
Ah, `Operator Mono` ... 
Trying to reproduce...

![image](https://user-images.githubusercontent.com/16012374/218257393-38b6e262-8f2e-4ff3-8098-624750caa369.png)

Now examining the names

![image](https://user-images.githubusercontent.com/16012374/218257483-aa394eb8-faca-4279-a2e2-71e31915e1c7.png)

Zoom into the family names:

```
 Family                                                  | | Subfamily                      | | Typogr. Family                           | | Typogr. Subfamily
 ------------------------------------------------------- |-| ------------------------------ |-| ---------------------------------------- |-| ----------------------------------------
 OperatorMono Book Nerd Font                             | | Italic                         | |                                          | |
 OperatorMono Book Nerd Font                             | | Regular                        | |                                          | |
 OperatorMonoExtra Nerd Font Light                       | | Italic                         | | OperatorMonoExtra Nerd Font              | | Light Italic
 OperatorMonoExtra Nerd Font Light                       | | Regular                        | | OperatorMonoExtra Nerd Font              | | Light
 OperatorMono Nerd Font                                  | | Bold Italic                    | |                                          | |
 OperatorMono Nerd Font                                  | | Bold                           | |                                          | |
 OperatorMono Nerd Font Light                            | | Italic                         | | OperatorMono Nerd Font                   | | Light Italic
 OperatorMono Nerd Font Light                            | | Regular                        | | OperatorMono Nerd Font                   | | Light
 OperatorMono Nerd Font Medium                           | | Italic                         | | OperatorMono Nerd Font                   | | Medium Italic
 OperatorMono Nerd Font Medium                           | | Regular                        | | OperatorMono Nerd Font                   | | Medium
```

Ok, I'll examine.
Thanks for reporting!

**Comment by Finii** [comment]
Should be fixed with 3.0.0
